cpp/include/cudf/detail/utilities/stream_pool.hpp

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/utilities/span.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+
+namespace cudf::detail {
+
+/**
+ * @brief Acquire a set of `cuda_stream_view` objects and synchronize them to an event on another
+ * stream.
+ *
+ * By default an underlying `rmm::cuda_stream_pool` is used to obtain the streams. The only other
+ * implementation at present is a debugging version that always returns the stream returned by
+ * `cudf::get_default_stream()`. To use this debugging version, set the environment variable
+ * `LIBCUDF_USE_DEBUG_STREAM_POOL`.
+ *
+ * Example usage:
+ * @code{.cpp}
+ * auto stream = cudf::get_default_stream();
+ * auto const num_streams = 2;
+ * // do work on stream
+ * // allocate streams and wait for an event on stream before executing on any of streams
+ * auto streams = cudf::detail::fork_streams(stream, num_streams);
+ * // do work on streams[0] and streams[1]
+ * // wait for event on streams before continuing to do work on stream
+ * cudf::detail::join_streams(streams, stream);
+ * @endcode
+ *
+ * @param stream Stream to synchronize the returned streams to, usually the default stream.
+ * @param count The number of stream views to return.
+ * @return Vector containing `count` stream views.
+ */
+std::vector<rmm::cuda_stream_view> fork_streams(rmm::cuda_stream_view stream, uint32_t count);
```

Comment:
should this be `fork_stream`?

Reply:
I guess that would be consistent with the input

Reply:
Why singular? It forks multiple streams from a single input. In the context of unix "fork", this is like calling `fork` multiple times. 

Plus, plural is consistent with `join_streams`

Reply:
Forking from the singular stream that is passed in as input? Joining multiple streams back to the singular stream. :shrug: I'm bad at naming so I'm happy either way.

Reply:
Interesting. My interpretation was that the single stream is forked into multiple streams. Hence singular.
I'd be okay with `fork_streams` as well.

Reply:
At first glance, I see this the same way as @etseidl and @vuule. It seems like you fork one stream to make multiple streams, and join multiple streams to make one stream. Hence `fork_stream` is singular, `join_streams` is plural based on their inputs. I'm not sure that the parallel to UNIX `fork` is strong enough to interpret it as requiring a plural. But I also see that it sounds nice to have both be plural.

Reply:
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent.  -->
**💡 suggestion:** ‏ The disagreement is really about whether we are naming the functions in reference to their input or their output. I have a suggestion that may resolve this difference of opinions: let the input and output types (and plurality) speak for themselves. Having `stream` or `streams` in the name is not necessary, since they are part of the type signature. So just name the functions like so:

```cpp
std::vector<rmm::cuda_stream_view> fork(rmm::cuda_stream_view stream, uint32_t count);
void join(host_span<rmm::cuda_stream_view> streams, rmm::cuda_stream_view stream);
```

<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities.  -->
**💭 thought (non-blocking):** ‏ I can also envision another version of `join` that doesn't take the singular output stream as input, and instead just chooses one stream to synchronize all the others to (e.g. the first in the span):

```cpp
rmm::cuda_stream_view join(host_span<rmm::cuda_stream_view> streams);
```

Reply:
Ahhh, missed this comment. `<luke>`They're coming too fast!`</luke>`

I'm ok with the renaming. I'd prefer to pass on the other `join`s and address them in the follow-on PRs.

Reply:
@bdice @vuule I think this is the last outstanding issue from this pass. Do you agree to the compromise naming suggestion?

Reply:
I have reservations about the name `cudf::detail::join` because it sounds like a dataframe join and not something with streams. If we can keep “stream” somewhere in the name (or namespace), I don’t care if it’s singular or plural.

Reply:
So let it be written, so shall it be done 😀